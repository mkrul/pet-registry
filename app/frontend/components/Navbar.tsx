import React from "react";
import { Link } from "react-router-dom";
import NavLink from "./shared/NavLink";
import SearchBar from "./shared/SearchBar";

const NavBar = () => {
  return (
    <div>
      {/* Main Navbar */}
      <div className="navbar bg-base-100">
        <div className="flex-1">
          <Link to="/" className="btn btn-ghost text-xl hover:bg-transparent">
            National Lost Pet Registry
          </Link>
        </div>

        {/* Navigation for larger screens */}
        <div className="hidden md:flex flex-none">
          <div className="align-right hidden lg:flex">
            <SearchBar />
          </div>
          <ul className="menu menu-horizontal px-1">
            <li>
              <NavLink linkTo="/reports/new">Report a Lost Pet</NavLink>
            </li>
            <li>
              <NavLink linkTo="#">About</NavLink>
            </li>
            <li>
              <NavLink linkTo="#">Contact</NavLink>
            </li>
          </ul>
        </div>

        {/* Profile Dropdown */}
        <div className="flex-none gap-2">
          <div className="dropdown dropdown-end">
            <div
              tabIndex={0}
              role="button"
              className="btn btn-ghost btn-circle avatar"
            >
              <div className="w-10 rounded-full">
                <img
                  alt="Tailwind CSS Navbar component"
                  src="https://img.daisyui.com/images/stock/photo-1534528741775-53994a69daeb.webp"
                />
              </div>
            </div>
            <ul
              tabIndex={0}
              className="menu menu-sm dropdown-content bg-base-100 rounded-box z-[1] mt-3 w-52 p-2 shadow"
            >
              <li>
                <NavLink linkTo="#">My Reports</NavLink>
              </li>
              <li>
                <NavLink linkTo="#">My Pets</NavLink>
              </li>
              <li>
                <NavLink linkTo="#">
                  Profile <span className="badge">New</span>
                </NavLink>
              </li>
              <li>
                <NavLink linkTo="#">Settings</NavLink>
              </li>
              <li>
                <NavLink linkTo="#">Logout</NavLink>
              </li>
            </ul>
          </div>
        </div>
      </div>

      {/* Navigation for smaller devices */}
      <div className="md:hidden w-full px-4 mt-2 mb-5 text-center">
        <ul className="flex justify-center space-x-6">
          <li>
            <NavLink linkTo="/reports/new">Report a Lost Pet</NavLink>
          </li>
          <li>
            <NavLink linkTo="#">About</NavLink>
          </li>
          <li>
            <NavLink linkTo="#">Contact</NavLink>
          </li>
        </ul>
      </div>

      {/* Full-width SearchBar for small screens */}
      <div className="lg:hidden w-full px-4 mt-3 mb-5">
        <SearchBar />
      </div>
    </div>
  );
};

export default NavBar;
