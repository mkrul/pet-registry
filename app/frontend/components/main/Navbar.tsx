import SearchBar from "../shared/SearchBar";

import DesktopMenu from "../main/DesktopMenu";
import MobileMenu from "./MobileMenu";

const NavBar = () => {
  return (
    <div>
      <DesktopMenu />

      <MobileMenu />

      <div className="lg:hidden w-full px-4 mt-2">
        <SearchBar />
      </div>
    </div>
  );
};

export default NavBar;
