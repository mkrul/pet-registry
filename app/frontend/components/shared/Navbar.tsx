import { Link } from "react-router-dom";
import ProfileDropdown from "../main/ProfileDropdown";
import NavLink from "./NavLink";

const NavBar = () => {
  return (
    <div>
      {/* Main Navbar */}
      <div className="navbar bg-base-100">
        <div className="flex-1">
          <Link to="/" className="btn btn-ghost text-xl hover:bg-transparent">
            National Lost Pet Registry
          </Link>
        </div>

        {/* Navigation for larger screens */}
        <div className="hidden md:flex flex-none">
          <ul className="menu menu-horizontal px-1">
            <li>
              <NavLink linkTo="/reports/new">Report a Lost Pet</NavLink>
            </li>
            <li>
              <NavLink linkTo="/reports">Search</NavLink>
            </li>
            <li>
              <NavLink linkTo="#">About</NavLink>
            </li>
            <li>
              <NavLink linkTo="#">Contact</NavLink>
            </li>
          </ul>
        </div>

        <ProfileDropdown />
      </div>

      {/* Navigation for smaller devices */}
      <div className="md:hidden w-full px-4 mt-2 mb-5 text-center">
        <ul className="flex justify-center space-x-6">
          <li>
            <NavLink linkTo="/reports/new">Report a Lost Pet</NavLink>
          </li>
          <li>
            <NavLink linkTo="/reports">Search</NavLink>
          </li>
          <li>
            <NavLink linkTo="#">About</NavLink>
          </li>
          <li>
            <NavLink linkTo="#">Contact</NavLink>
          </li>
        </ul>
      </div>
    </div>
  );
};

export default NavBar;
