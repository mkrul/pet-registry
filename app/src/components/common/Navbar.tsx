import { Link, useLocation } from "react-router-dom";
import { useState, useRef, useEffect } from "react";
import ProfileDropdown from "../main/ProfileDropdown";
import NavLink from "./NavLink";

const NavBar = () => {
  const [isMenuOpen, setIsMenuOpen] = useState(false);
  const menuRef = useRef<HTMLDivElement>(null);
  const location = useLocation();

  const handleHomeClick = (e: React.MouseEvent) => {
    e.preventDefault();
    window.location.replace(window.location.origin);
  };

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (menuRef.current && !menuRef.current.contains(event.target as Node)) {
        setIsMenuOpen(false);
      }
    };

    if (isMenuOpen) {
      document.addEventListener("mousedown", handleClickOutside);
    }

    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, [isMenuOpen]);

  // Close menu when route changes
  useEffect(() => {
    setIsMenuOpen(false);
  }, [location.pathname]);

  const handleMenuToggle = () => {
    setIsMenuOpen(!isMenuOpen);
  };

  return (
    <nav className="bg-page">
      <div data-testid="navbar">
        {/* Main Navbar */}
        <div className="navbar">
          {/* Hamburger Menu for Mobile/Tablet */}
          <div className="flex-none sm:hidden">
            <div className="dropdown" ref={menuRef}>
              <button
                className="btn btn-ghost btn-circle"
                onClick={handleMenuToggle}
                data-testid="hamburger-menu-button"
                aria-label="Toggle navigation menu"
              >
                <svg
                  className="w-6 h-6"
                  fill="none"
                  stroke="currentColor"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    strokeLinecap="round"
                    strokeLinejoin="round"
                    strokeWidth={2}
                    d="M4 6h16M4 12h16M4 18h16"
                  />
                </svg>
              </button>
              <ul
                className={`menu menu-sm dropdown-content bg-white rounded-box z-[1] mt-3 w-52 p-2 shadow ${
                  isMenuOpen ? "block" : "hidden"
                }`}
                data-testid="mobile-menu"
              >
                <li className="hover:bg-base-200 rounded-lg transition-colors duration-200">
                  <NavLink linkTo="/reports/new" handler={() => setIsMenuOpen(false)}>Report</NavLink>
                </li>
                <li className="hover:bg-base-200 rounded-lg transition-colors duration-200">
                  <NavLink linkTo="/reports" handler={() => setIsMenuOpen(false)}>Search</NavLink>
                </li>
                <li className="hover:bg-base-200 rounded-lg transition-colors duration-200">
                  <NavLink linkTo="#" handler={() => setIsMenuOpen(false)}>About</NavLink>
                </li>
                <li className="hover:bg-base-200 rounded-lg transition-colors duration-200">
                  <NavLink linkTo="#" handler={() => setIsMenuOpen(false)}>Contact</NavLink>
                </li>
              </ul>
            </div>
          </div>

          <div className="flex-1">
            <Link
              to="/"
              onClick={handleHomeClick}
              className="btn btn-ghost text-xl hover:bg-transparent pl-1 sm:pl-4"
            >
              Lost Pets Registry
            </Link>
          </div>

          {/* Navigation for larger screens (850px+) */}
          <div className="hidden md:flex flex-none">
            <ul
              className="menu menu-horizontal [&>li>a]:px-[0.75rem]"
              data-testid="desktop-nav"
              aria-label="desktop navigation"
            >
              <li>
                <NavLink linkTo="/reports/new">Report a Lost Pet</NavLink>
              </li>
              <li>
                <NavLink linkTo="/reports">Search</NavLink>
              </li>
              <li>
                <NavLink linkTo="#">About</NavLink>
              </li>
              <li>
                <NavLink linkTo="#">Contact</NavLink>
              </li>
            </ul>
          </div>

          {/* Navigation for tablet sizes (640px-849px) */}
          <div className="hidden sm:flex md:hidden flex-none">
            <ul
              className="menu menu-horizontal [&>li>a]:px-[0.75rem]"
              data-testid="tablet-nav"
              aria-label="tablet navigation"
            >
              <li>
                <NavLink linkTo="/reports/new">Report</NavLink>
              </li>
              <li>
                <NavLink linkTo="/reports">Search</NavLink>
              </li>
              <li>
                <NavLink linkTo="#">About</NavLink>
              </li>
              <li>
                <NavLink linkTo="#">Contact</NavLink>
              </li>
            </ul>
          </div>

          <ProfileDropdown />
        </div>
      </div>
    </nav>
  );
};

export default NavBar;
