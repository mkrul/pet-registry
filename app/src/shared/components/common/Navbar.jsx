import { Link, useLocation } from "react-router-dom";
import { useState, useRef, useEffect } from "react";
import ProfileDropdown from "../layout/ProfileDropdown";
import NavLink from "./NavLink";
import { useSelector } from "react-redux";
const NavBar = () => {
  const [isMenuOpen, setIsMenuOpen] = useState(false);
  const menuRef = useRef(null);
  const location = useLocation();
  const isAuthenticated = useSelector((state) => !!state.auth.user);

  const handleHomeClick = (e) => {
    e.preventDefault();
    window.location.replace(window.location.origin);
  };

  useEffect(() => {
    const handleClickOutside = (event) => {
      if (menuRef.current && !menuRef.current.contains(event.target)) {
        setIsMenuOpen(false);
      }
    };

    if (isMenuOpen) {
      document.addEventListener("mousedown", handleClickOutside);
    }

    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, [isMenuOpen]);

  // Close menu when route changes
  useEffect(() => {
    setIsMenuOpen(false);
  }, [location.pathname]);

  const handleMenuToggle = () => {
    setIsMenuOpen(!isMenuOpen);
  };

  return (
    <nav className="bg-page">
      <div data-testid="navbar">
        {/* Main Navbar */}
        <div className="navbar">
          {/* Hamburger Menu for Mobile/Tablet */}
          <div className="flex-none sm:hidden">
            <div className="dropdown" ref={menuRef}>
              <button
                className="btn btn-ghost btn-circle"
                onClick={handleMenuToggle}
                data-testid="hamburger-menu-button"
                aria-label="Toggle navigation menu"
              >
                <svg
                  className="w-6 h-6"
                  fill="none"
                  stroke="currentColor"
                  viewBox="0 0 24 24"
                  xmlns="http:/www.w3.org/2000/svg"
                >
                  <path
                    strokeLinecap="round"
                    strokeLinejoin="round"
                    strokeWidth={2}
                    d="M4 6h16M4 12h16M4 18h16"
                  />
                </svg>
              </button>
              <ul
                className={`menu menu-sm dropdown-content bg-white rounded-box z-[1] mt-3 w-52 p-2 shadow ${
                  isMenuOpen ? "block" : "hidden"
                }`}
                data-testid="mobile-menu"
              >
                {isAuthenticated && (
                  <li className="hover:bg-base-200 rounded-lg transition-colors duration-200">
                    <NavLink linkTo="/dashboard/reports?action=create" handler={() => setIsMenuOpen(false)}>Report a Lost Pet</NavLink>
                  </li>
                )}
                <li className="hover:bg-base-200 rounded-lg transition-colors duration-200">
                  <NavLink linkTo="/reports" handler={() => setIsMenuOpen(false)}>Search</NavLink>
                </li>
                <li className="hover:bg-base-200 rounded-lg transition-colors duration-200">
                  <NavLink linkTo="/about" handler={() => setIsMenuOpen(false)}>About</NavLink>
                </li>
              </ul>
            </div>
          </div>

          <div className="flex-1">
            <Link
              to="/"
              onClick={handleHomeClick}
              className="btn btn-ghost text-xl hover:bg-transparent pl-1 sm:pl-4"
            >
              Lost Pets Registry
            </Link>
          </div>

          {/* Navigation for screens 640px and larger */}
          <div className="hidden sm:flex flex-none">
            <ul
              className="menu menu-horizontal [&>li>a]:px-[0.875rem]"
              data-testid="desktop-nav"
              aria-label="desktop navigation"
            >
              {isAuthenticated && (
                <li>
                  <NavLink linkTo="/dashboard/reports?action=create">Report a Lost Pet</NavLink>
                </li>
              )}
              <li>
                <NavLink linkTo="/reports">Search</NavLink>
              </li>
              <li>
                <NavLink linkTo="/about">About</NavLink>
              </li>
              {!isAuthenticated && (
                <li>
                  <NavLink linkTo="/login" data-testid="nav-link-login">Log In</NavLink>
                </li>
              )}
            </ul>
          </div>

          <ProfileDropdown />
        </div>
      </div>
    </nav>
  );
};

export default NavBar;
