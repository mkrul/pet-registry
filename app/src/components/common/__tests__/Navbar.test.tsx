import { describe, it, expect, vi, beforeAll, afterAll, beforeEach } from "vitest";
import { render, screen, within } from "@testing-library/react";
import "@testing-library/jest-dom/vitest";
import { MemoryRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import Navbar from "../Navbar";

// Mock the ProfileDropdown component
vi.mock("../../main/ProfileDropdown", () => ({
  default: () => <div data-testid="profile-dropdown">Profile Dropdown</div>
}));

// Mock NavLink component
vi.mock("../NavLink", () => ({
  default: ({ children, linkTo }: { children: React.ReactNode; linkTo: string }) => (
    <a href={linkTo} data-testid="nav-link">
      {children}
    </a>
  )
}));

describe("Navbar", () => {
  const originalLocation = window.location;
  const mockReplace = vi.fn();

  beforeAll(() => {
    Object.defineProperty(window, "location", {
      configurable: true,
      value: { replace: mockReplace, origin: "http://localhost" }
    });
  });

  afterAll(() => {
    Object.defineProperty(window, "location", {
      configurable: true,
      value: originalLocation
    });
  });

  beforeEach(() => {
    vi.clearAllMocks();
  });

  const renderNavbar = () => {
    return render(
      <MemoryRouter>
        <Navbar />
      </MemoryRouter>
    );
  };

  describe("Desktop Navigation", () => {
    it("renders logo with correct link and text", () => {
      renderNavbar();
      const logo = screen.getByText("Lost Pet Registry");
      expect(logo).toBeDefined();
      expect(logo).toHaveClass("btn", "btn-ghost", "text-xl", "hover:bg-transparent");
    });

    it("renders desktop navigation menu with correct classes", () => {
      renderNavbar();
      const desktopNav = screen.getByRole("list", {
        hidden: false,
        name: /desktop navigation/i
      });
      expect(desktopNav).toHaveClass("menu", "menu-horizontal", "px-1");
    });

    it("renders all desktop navigation links", () => {
      renderNavbar();
      const desktopContainer = screen.getByTestId("desktop-nav");
      const desktopLinks = within(desktopContainer).getAllByTestId("nav-link");
      expect(desktopLinks).toHaveLength(4);
    });
  });

  describe("Mobile Navigation", () => {
    it("renders mobile navigation menu with correct classes", () => {
      renderNavbar();
      const mobileNav = screen.getByTestId("mobile-nav");
      expect(mobileNav.parentElement).toHaveClass("md:hidden");
    });

    it("renders all mobile navigation links", () => {
      renderNavbar();
      const mobileContainer = screen.getByTestId("mobile-nav");
      const mobileLinks = within(mobileContainer).getAllByTestId("nav-link");
      expect(mobileLinks).toHaveLength(4);
    });
  });

  describe("Navigation Links", () => {
    const expectedLinks = [
      { text: "Report a Lost Pet", path: "/reports/new" },
      { text: "Search", path: "/reports" },
      { text: "About", path: "#" },
      { text: "Contact", path: "#" }
    ];

    it.each(expectedLinks)("renders $text link with correct href", ({ text, path }) => {
      renderNavbar();
      const links = screen.getAllByText(text);
      links.forEach(link => {
        expect(link.closest("a")).toHaveAttribute("href", path);
      });
    });
  });

  describe("Interactions", () => {
    it("handles home link click by replacing location", async () => {
      renderNavbar();
      const homeLink = screen.getByText("Lost Pet Registry");
      await userEvent.click(homeLink);
      expect(mockReplace).toHaveBeenCalledWith("http://localhost");
    });

    it("prevents default behavior on home link click", async () => {
      renderNavbar();
      const homeLink = screen.getByText("Lost Pet Registry");
      const preventDefault = vi.fn();
      homeLink.dispatchEvent(new MouseEvent("click", { bubbles: true, cancelable: true }));
      expect(preventDefault).not.toHaveBeenCalled();
    });
  });

  describe("Components", () => {
    it("renders ProfileDropdown component", () => {
      renderNavbar();
      expect(screen.getByTestId("profile-dropdown")).toBeDefined();
    });
  });

  describe("Snapshots", () => {
    it("matches component snapshot", () => {
      const { container } = renderNavbar();
      expect(container).toMatchSnapshot();
    });
  });
});
