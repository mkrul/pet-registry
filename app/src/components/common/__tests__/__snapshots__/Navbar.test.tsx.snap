// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navbar > Snapshots > matches component snapshot 1`] = `
<div>
  <div
    data-testid="navbar"
  >
    <div
      class="navbar bg-base-100"
    >
      <div
        class="flex-1"
      >
        <a
          class="btn btn-ghost text-xl hover:bg-transparent"
          href="/"
        >
          Lost Pet Registry
        </a>
      </div>
      <div
        class="hidden md:flex flex-none"
      >
        <ul
          aria-label="desktop navigation"
          class="menu menu-horizontal px-1"
          data-testid="desktop-nav"
        >
          <li>
            <a
              data-testid="nav-link"
              href="/reports/new"
            >
              Report a Lost Pet
            </a>
          </li>
          <li>
            <a
              data-testid="nav-link"
              href="/reports"
            >
              Search
            </a>
          </li>
          <li>
            <a
              data-testid="nav-link"
              href="#"
            >
              About
            </a>
          </li>
          <li>
            <a
              data-testid="nav-link"
              href="#"
            >
              Contact
            </a>
          </li>
        </ul>
      </div>
      <div
        data-testid="profile-dropdown"
      >
        Profile Dropdown
      </div>
    </div>
    <div
      class="md:hidden w-full px-4 mt-2 mb-5 text-center"
    >
      <ul
        aria-label="mobile navigation"
        class="flex justify-center space-x-6"
        data-testid="mobile-nav"
      >
        <li>
          <a
            data-testid="nav-link"
            href="/reports/new"
          >
            Report a Lost Pet
          </a>
        </li>
        <li>
          <a
            data-testid="nav-link"
            href="/reports"
          >
            Search
          </a>
        </li>
        <li>
          <a
            data-testid="nav-link"
            href="#"
          >
            About
          </a>
        </li>
        <li>
          <a
            data-testid="nav-link"
            href="#"
          >
            Contact
          </a>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
